examples/bcc/xdp/xdp_drop.go

```diff
@@ -0,0 +1,180 @@
+    struct iphdr *iph = data + nh_off;
+
+    if ((void*)&iph[1] > data_end)
+        return 0;
+    return iph->protocol;
+}
+
+static inline int parse_ipv6(void *data, u64 nh_off, void *data_end) {
+    struct ipv6hdr *ip6h = data + nh_off;
+
+    if ((void*)&ip6h[1] > data_end)
+        return 0;
+    return ip6h->nexthdr;
+}
+
+int xdp_prog1(struct CTXTYPE *ctx) {
+
+    void* data_end = (void*)(long)ctx->data_end;
+    void* data = (void*)(long)ctx->data;
+
+    struct ethhdr *eth = data;
+
+    // drop packets
+    int rc = RETURNCODE; // let pass XDP_PASS or redirect to tx via XDP_TX
+    long *value;
+    uint16_t h_proto;
+    uint64_t nh_off = 0;
+    uint32_t index;
+
+    nh_off = sizeof(*eth);
+
+    if (data + nh_off  > data_end)
+        return rc;
+
+    h_proto = eth->h_proto;
+
+    if (h_proto == htons(ETH_P_8021Q) || h_proto == htons(ETH_P_8021AD)) {
+        struct vlan_hdr *vhdr;
+
+        vhdr = data + nh_off;
+        nh_off += sizeof(struct vlan_hdr);
+        if (data + nh_off > data_end)
+            return rc;
+            h_proto = vhdr->h_vlan_encapsulated_proto;
+    }
+    if (h_proto == htons(ETH_P_8021Q) || h_proto == htons(ETH_P_8021AD)) {
+        struct vlan_hdr *vhdr;
+
+        vhdr = data + nh_off;
+        nh_off += sizeof(struct vlan_hdr);
+        if (data + nh_off > data_end)
+            return rc;
+            h_proto = vhdr->h_vlan_encapsulated_proto;
+    }
+
+    if (h_proto == htons(ETH_P_IP))
+        index = parse_ipv4(data, nh_off, data_end);
+    else if (h_proto == htons(ETH_P_IPV6))
+       index = parse_ipv6(data, nh_off, data_end);
+    else
+        index = 0;
+
+    value = dropcnt.lookup(&index);
+    if (value)
+        *value += 1;
+
+    return rc;
+}
+`
+
+func usage() {
+	fmt.Printf("Usage: %v <ifdev>\n", os.Args[0])
+	fmt.Printf("e.g.: %v eth0\n", os.Args[0])
+	os.Exit(1)
+}
+
+func main() {
+	var device string
+
+	if len(os.Args) != 2 {
+		usage()
+	}
+
+	device = os.Args[1]
+
+	ret := "XDP_DROP"
+	ctxtype := "xdp_md"
+
+	module := bpf.NewModule(source, []string{
+		"-w",
+		"-DRETURNCODE=" + ret,
+		"-DCTXTYPE=" + ctxtype,
+	})
+	defer module.Close()
+
+	fn, err := module.Load("xdp_prog1", C.BPF_PROG_TYPE_XDP)
+
+	err = module.AttachXDP(device, fn)
+
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	defer func() {
+		if err := module.RemoveXDP(device); err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to remove XDP from %s: %v\n", device, err)
+		}
+	}()
+
+	fmt.Println("Dropping packets, hit CTRL+C to stop")
+
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, os.Kill)
+
+	dropcnt := bpf.NewTable(module.TableId("dropcnt"), module)
+
+	<-sig
+
+	fmt.Printf("\n{IP protocol-number}: {total dropped pkts}\n")
+	for entry := range dropcnt.Iter() {
+		var key, value uint64
+		var err error
+
+		key, err = strconv.ParseUint(entry.Key, 0, 32)
+		if err != nil {
+			continue
+		}
+
+		value, err = strconv.ParseUint(entry.Value, 0, 64)
```

Comment:
Is it the correct way to get the value for each CPU?
This page suggests the lookup returns an array with the values for each CPU:

https://www.mail-archive.com/netdev@vger.kernel.org/msg95249.html